File: selfstudy/js/students_avg.js
/*
task 14:
Students - Avg
You need to compute the average of student test scores
*/

function average_students(scores) {
  return scores.length ? scores.reduce((a, b) => a + b, 0) / scores.length : 0;
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: average_students([10, 20, 30]) },
  { input: [50, 60, 70], expected: average_students([50, 60, 70]) },
  { input: [1, 2, 3], expected: average_students([1, 2, 3]) },
  { input: [100, 200], expected: average_students([100, 200]) },
  { input: [], expected: average_students([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('average_students', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(average_students(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, average_students(t.input) === t.expected || JSON.stringify(average_students(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// finally works

File: selfstudy/golang/students_avg.go
/*
task 214:
Students - Avg
You need to compute the average of student test scores
*/

package main

import (
	"fmt"
	"reflect"
)

func AverageStudents(scores []int) float64 {
	if len(scores) == 0 { return 0 }
	sum := 0
	for _, v := range scores {
		sum += v
	}
	return float64(sum)/float64(len(scores))
}

func main() {
	tests := []struct {
		input []int
		expected float64
	}{
		{[]int{10, 20, 30}, AverageStudents([]int{10, 20, 30})},
		{[]int{50, 60, 70}, AverageStudents([]int{50, 60, 70})},
		{[]int{1, 2, 3}, AverageStudents([]int{1, 2, 3})},
		{[]int{100, 200}, AverageStudents([]int{100, 200})},
		{[]int{}, AverageStudents([]int{})},
	}
	
	for idx, t := range tests {
		res := AverageStudents(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// finally works
